R/consume-data.R

```diff
@@ -100,39 +100,30 @@ get_via_csv <- function(ss, ws = 1, ...) {
 #'
 #' @export
 get_via_lf <- function(ss, ws = 1) {
-
+  
   stopifnot(ss %>% inherits("googlesheet"))
   
   this_ws <- get_ws(ss, ws)
   req <- gsheets_GET(this_ws$listfeed)
-  row_data <- req$content %>% lfilt("entry")
-  
-  ## when I switch to xml2 and/or stop converting feed via xmlToList(), I can
-  ## dramatically rationalize the separation of boilerplate from content
-  ## TO DO: experiment with empty header cells, header cells with
-  ## space or other non-alphanumeric characters
-  ## https://developers.google.com/google-apps/spreadsheets/#working_with_list-based_feeds
-  
-  component_names <- row_data[[1]] %>% names()
-  boilerplate_names <-
-    c("id", "updated", "category", "title", "content", "link")
-  last_boilerplate <-
-    which(!(component_names %in% boilerplate_names)) %>%
-    min() %>%
-    `-`(1)
-  drop_boilerplate_vars <- -1 * seq_len(last_boilerplate)
-  var_names <- component_names[drop_boilerplate_vars]
-  
-  ## used below to handle empty cells: replace NULL with NA
-  null_NA_character <- function(x) if(is.null(x)) NA_character_ else x
-  
-  row_data %>%
-    unname() %>% # drop stupid and ubiquitous 'entry' names
-    plyr::llply(`[`, drop_boilerplate_vars) %>%
-    plyr::llply(function(x) unlist(lapply(x, null_NA_character))) %>%
-    do.call("rbind", .) %>%
-    plyr::alply(2, type.convert, as.is = TRUE, .dims = TRUE)  %>%
-    ## get rid of attributes that are non-standard for tbl_dfs or data.frames
+  
+  ns <- xml2::xml_ns_rename(xml2::xml_ns(req$content), d1 = "feed")
+  
+  var_names <- req$content %>% 
+    xml2::xml_find_all("//feed:entry//gsx:*", ns) %>% 
```

Comment:
For example, here's how you could get the variable names from the first row = entry:

```
req$content %>% 
    xml2::xml_find_all("(//feed:entry)[1]", ns) %>%
    xml2::xml_find_all(".//gsx:*", ns) %>%
    xml2::xml_name()
```